README.md
# mongodb-replica-maintainer #

A server and client for adding/removing dynamic mongodb replica members.

### Usage ###

Example systemd service files are included in the `systemd` folder. You'll need to modify
the `--server` argument in the client service to point to the server.

```
# Client
$ node client.js --server=ws://127.0.0.1:27018
```

```
# Server
$ node server.js
```

### SkyDNS ###

If you want to use SkyAPI/SkyDNS for the client to discover the server, then use the
`--skyapi` flag when starting the server. The server will advertise itself as
`mongodb-replica-maintainer` so for the client you would run:
```
node client.js --server=srv://mongodb-replica-maintainer.services.domain
```
where `domain` is the SkyDNS suffix you're using.

By [James Hartig](https://github.com/fastest963/)


CHANGELOG.md
## Changelog ##

### 0.5.2 ###
* Upgrade to newest skyprovider

### 0.5.0 ###
* Switch to using modulelog

### 0.4.1 ###
* Fixed srv lookup not happening on reconnect
* Updated mongodb to 2.0.41
* Added mongodb options to keep retrying on disconnect


client.js
var flags = require('flags'),
    WebSocket = require('ws'),
    Log = require('modulelog')('mongodb-replica-maintainer'),
    pendingConnect = null,
    pendingReconnect = null,
    pendingAdd = null,
    pingInterval = null,
    server = '',
    ourHostname = '';

flags.defineString('server', '', 'maintainer server endpoint');
flags.defineString('host', '', 'mongodb host string or "" to rely on server');
flags.defineBoolean('hidden', false, 'set the replica hidden option');
flags.defineInteger('ping', 15000, 'ms between pings (2 pings means failed)');
flags.defineInteger('reconnect', 5000, 'ms before reconnecting');
flags.defineInteger('priority', 0, 'this replica\'s priority');
flags.defineInteger('votes', 0, 'this replica\'s votes');
flags.defineString('logger', 'default', 'the class to use for logging');
flags.defineString('log-level', 'info', 'the log level');
flags.parse();
server = flags.get('server');

Log.setClass(flags.get('logger'));
Log.setLevel(flags.get('log-level'));

function reconnect() {
    if (pendingReconnect || !flags.get('reconnect')) {
        return;
    }
    //if we lose the connection, try to reconnect in x seconds
    pendingReconnect = setTimeout(resolveAndConnect, Math.max(100, flags.get('reconnect')));
}

function resolveAndConnect() {
    if (server.indexOf('srv://') === 0) {
        var srv = require('srvclient'),
            hostname = server.substr(6);
        Log.info('Resolving srv address', {hostname: hostname});
        srv.getTarget(hostname, function(err, target) {
            if (err) {
                Log.error('Failed to resolve srv hostname', {hostname: hostname, error: err});
                setTimeout(resolveAndConnect, Math.max(100, flags.get('reconnect')));
                return;
            }
            Log.debug('Resolving srv target', {target: target, hostname: hostname});
            target.resolve(function(err, addr) {
                if (err) {
                    Log.error('Failed to resolve srv target', {target: target, hostname: hostname, error: err});
                    setTimeout(resolveAndConnect, Math.max(100, flags.get('reconnect')));
                    return;
                }
                connect('ws://' + [addr, target.port].join(':'));
            });
        });
    } else {
        connect(server);
    }
}

function sendAdd(ws) {
    clearTimeout(pendingAdd);
    pendingAdd = null;
    if (ws.readyState !== WebSocket.OPEN) {
        return;
    }
    var host = flags.get('host') || undefined,
        hidden = flags.get('hidden'),
        pri = flags.get('priority'),
        votes = flags.get('votes');
    Log.debug('Sending add', {host: host, hidden: hidden, priority: pri, votes: votes});
    ws.send(JSON.stringify({
        cmd: 'add',
        host: host,
        hidden: hidden,
        priority: pri,
        votes: votes
    }));
    pendingAdd = setTimeout(function() {
        Log.error("Timed out waiting for response to add", {url: ws.url});
        sendAdd(ws);
    }, 15000);
}

function startPing(ws) {
    clearInterval(pingInterval);
    if (!flags.get('ping')) {
        return;
    }
    var interval = Math.max(1000, flags.get('ping'));
    pingInterval = setInterval(function() {
        if (ws.readyState !== WebSocket.OPEN) {
            return;
        }
        ws.ping();
        if (Date.now() - ws.lastPong >= (2 * interval)) {
            Log.error("Timed out waiting for pong", {url: ws.url});
            ws.close();
        }
    }, interval);
}

function connect(address) {
    clearTimeout(pendingReconnect);
    pendingReconnect = null;
    clearTimeout(pendingConnect);
    pendingConnect = null;

    if (address.indexOf('://') === -1) {
        address = 'ws://' + address;
    }

    Log.debug("Connecting to ws", {address: address});

    var ws = new WebSocket(address);
    ws.on('open', function() {
        Log.info("Connected to ws", {address: address});
        clearTimeout(pendingConnect);
        pendingConnect = null;
        //since we just connected we got a "pong"
        ws.lastPong = Date.now();
        sendAdd(ws);
    });
    ws.on('message', function(message) {
        Log.debug("Received message", {message: message});
        var result;
        try {
            result = JSON.parse(message);
        } catch (e) {
            Log.error("Received invalid json", {message: message, error: e});
            ws.close();
            return;
        }
        if (result.hasOwnProperty('added')) {
            clearTimeout(pendingAdd);
            pendingAdd = null;
            if (!result.success) {
                Log.error("Failed to add", {error: result.error});
                pendingAdd = setTimeout(function() {
                    sendAdd(ws);
                }, (Math.random() * 5000) + 10000); //randomly between 10-15 seconds
            }
            Log.info("Added", {hostname: result.added});
            ourHostname = result.added;
            //since we just received a message we got a "pong"
            ws.lastPong = Date.now();
            startPing(ws);
            return;
        }
        if (result.hasOwnProperty('removed') && result.removed === ourHostname && pendingAdd === null) {
            Log.error("Removed from replica set", {hostname: result.removed});
            pendingAdd = setTimeout(function() {
                sendAdd(ws);
            }, (Math.random() * 5000) + 10000); //randomly between 10-15 seconds
        }
    });
    ws.on('error', function(err) {
        Log.error("Error from ws", {error: err});
        clearInterval(pingInterval);
        pingInterval = null;
        clearInterval(pendingAdd);
        pendingAdd = null;
        reconnect();
    });
    ws.on('pong', function() {
        Log.debug("Received pong");
        ws.lastPong = Date.now();
    });
    ws.on('close', function() {
        Log.error("Disconnected from ws", {address: address});
        clearInterval(pingInterval);
        pingInterval = null;
        clearInterval(pendingAdd);
        pendingAdd = null;
        reconnect();
    });

    pendingConnect = setTimeout(function() {
        Log.error("Timed out waiting to connect to ws", {address: address});
        ws.terminate();
        reconnect();
    }, 15000);
}

resolveAndConnect();


server.js
var flags = require('flags'),
    MongoClient = require('mongodb').MongoClient,
    SkyProvider = require('skyprovider'),
    WebSocket = require('ws'),
    WebSocketServer = require('ws').Server,
    Log = require('modulelog')('mongodb-replica-maintainer'),
    membersJustAdded = {},
    gracePeriod = 15 * 60 * 1000, //give clients 15 minutes to start up before trying to remove
    allClients = [];

flags.defineString('ip', '0.0.0.0', 'listen ip');
flags.defineInteger('port', 27018, 'listen port');
flags.defineString('mongo', 'mongodb://127.0.0.1:27017', 'mongodb connect string');
flags.defineBoolean('clean', true, 'clean up dead replicas');
flags.defineString('skyapi', '', 'skyapi address to advertise to');
flags.defineString('logger', 'default', 'the class to use for logging');
flags.defineString('log-level', 'info', 'the log level');
flags.parse();

Log.setClass(flags.get('logger'));
Log.setLevel(flags.get('log-level'));

function cleanupDeadReplicas(db) {
    var adminDB = db.admin();
    adminDB.command({replSetGetStatus: 1}, function(err, status) {
        if (!status || !status.ok || !status.members) {
            if (!status) {
                Log.error('Error calling replSetGetStatus', {error: err});
            } else {
                Log.error('Error calling replSetGetStatus', {status: status});
            }
            return;
        }
        var i = 0,
            now = Date.now(),
            idsToRemove = [],
            member;
        for (; i < status.members.length; i++) {
            member = status.members[i];
            //state of 1 means primary
            if (member.state !== 1 && !member.health && status.date - member.lastHeartbeatRecv > (3600 * 1000)) {
                if (membersJustAdded.hasOwnProperty(member.name)) {
                    if (now - membersJustAdded[member.name] < gracePeriod) {
                        continue;
                    }
                    delete membersJustAdded[member.name];
                }
                idsToRemove.push(status.members[i]._id);
            }
        }
        Log.debug('Removing dead replicas', {count: idsToRemove.length});
        if (idsToRemove.length < 1) {
            return;
        }
        db.collection('system.replset').findOne({}, function(err, doc) {
            if (err) {
                cb(err);
                return;
            }
            var changed = false;
            for (i = 0; i < doc.members.length; i++) {
                if (idsToRemove.indexOf(doc.members[i]._id) !== -1) {
                    Log.info('Removing dead replica', {host: doc.members[i].host});
                    broadcastRemoval(doc.members[i].host);
                    doc.members.splice(i, 1);
                    i--;
                    changed = true;
                }
            }
            if (!changed) {
                return;
            }
            doc.version++;
            db.admin().command({replSetReconfig: doc}, function(err, res) {
                if (err) {
                    Log.error('Error removing dead replicas', {error: err});
                }
            });
        });
    });
}

function broadcastRemoval(host) {
    var msg = JSON.stringify({removed: host});
    for (var i = 0; i < allClients.length; i++) {
        try {
            allClients[i].send(msg);
        } catch (e) {
            Log.error('Error broadcasting removal', {error: e});
            allClients.splice(i, 1);
            i--;
        }
    }
}

function addNewReplicaMember(db, host, hidden, priority, votes, cb) {
    db.collection('system.replset').findOne({}, function(err, doc) {
        if (err) {
            cb(err);
            return;
        }
        if (!doc || !Array.isArray(doc.members)) {
            Log.error('invalid replset doc', {doc: doc});
            cb(new TypeError('Invalid members array in replset collection'));
            return;
        }
        if (host.indexOf(':') === -1) {
            host += ':27017';
        }
        var found = -1,
            highestID = 0,
            i = 0,
            changed = false;
        for (; i < doc.members.length; i++) {
            if (doc.members[i].host === host) {
                found = i;
                break;
            }
            highestID = Math.max(highestID, doc.members[i]._id);
        }
        if (found < 0) {
            highestID++;
            Log.debug('Adding new replica', {host: host, id: highestID});
            //priority 0 means it will never be a primary
            doc.members.push({_id: highestID, host: host, priority: priority, votes: votes, hidden: hidden});
            changed = true;
        } else {
            if (doc.members[found].hidden !== hidden) {
                doc.members[found].hidden = hidden;
                changed = true;
            }
            if (doc.members[found].priority !== priority) {
                doc.members[found].priority = priority;
                changed = true;
            }
            if (doc.members[found].votes !== votes) {
                doc.members[found].votes = votes;
                changed = true;
            }
            if (changed) {
                Log.debug('Updating existing replica', {host: host});
            }
        }
        if (!changed) {
            cb(null, host);
            return;
        }
        doc.version++;
        db.admin().command({replSetReconfig: doc}, function(err, res) {
            if (err) {
                cb(err);
                return;
            }
            membersJustAdded[host] = Date.now();
            cb(null, host);
        });
    });
}

function handleMessage(db, ws, message) {
    var ip = ws._socket.remoteAddress,
        command;
    try {
        command = JSON.parse(message);
    } catch (e) {
        Log.error('Received invalid json', {error: e, ip: ip});
        ws.close();
        return;
    }
    if (!command.cmd) {
        Log.error('Received message without cmd', {command: command, ip: ip});
        ws.send('{"success":false,"error":"no cmd"}');
        return;
    }
    switch (command.cmd) {
        case 'add':
            //todo: validate the host
            if (!command.host) {
                Log.debug('Assuming ip from sender', {ip: ip});
                command.host = ip;
            }
            if (command.hidden === undefined) {
                command.hidden = false;
            }
            if (command.priority === undefined) {
                command.priority = 0;
            }
            if (command.votes === undefined) {
                command.votes = 0;
            }
            addNewReplicaMember(db, command.host, command.hidden, command.priority, command.votes, function(err, hostAdded) {
                if (err) {
                    Log.error('Failed to add replica', {ip: ip, host: command.host, error: err});
                    ws.send('{"success":false,"error":"internal error"}');
                    return;
                }
                Log.info('Added new replica', {host: hostAdded, ip: ip});
                //make sure the socket is still open before trying to respond
                if (ws.readyState !== WebSocket.OPEN) {
                    return;
                }
                ws.send(JSON.stringify({added: hostAdded, success: true}));
                allClients.push(ws);
            });
            break;
        default:
            Log.error('Unknown command received', {command: command.cmd, ip: ip});
            ws.send('{"success":false,"error":"unknown cmd"}');
            break;
    }
}

function advertiseToSkyAPI() {
    var skyapiAddr = flags.get('skyapi'),
        port = flags.get('port');
    if (!skyapiAddr) {
        return;
    }
    if (skyapiAddr.indexOf('://') === -1) {
        skyapiAddr = 'ws://' + skyapiAddr;
    }
    if (skyapiAddr.indexOf('/provide') === -1) {
        skyapiAddr += '/provide';
    }
    Log.info('Advertising to skyapi', {skyapi: skyapiAddr, port: port});
    var provider = new SkyProvider(skyapiAddr);
    provider.provideService('mongodb-replica-maintainer', port);
}

var options = {},
    serverOptions = {
        autoReconnect: true,
        keepAlive: 5000, //wait 5 seconds before starting keep-alive
        connectTimeoutMS: 5000,
        noDelay: true
    };
options.db = {
    bufferMaxEntries: 0 //do not buffer any commands
};
options.server = {
    socketOptions: serverOptions,
    reconnectTries: 432000, //5 days in seconds
    reconnectInterval: 1000
};
options.replSet = {
    socketOptions: serverOptions,
    connectWithNoPrimary: false
};
MongoClient.connect(flags.get('mongo'), options, function(err, db) {
    if (err) {
        Log.error('Error connecting to mongo', {error: err});
        throw err;
    }
    var topology = db.s.topology,
        localDB = db.db('local'),
        numNoPrimaries = 0,
        server = null;

    Log.info('Connected to mongo', {addr: flags.get('mongo')});

    if (topology) {
        topology.on('ha', function(type, data) {
            if (type !== 'end') {
                return;
            }
            if (data.state.primary) {
                numNoPrimaries = 0;
                return;
            }
            if (!data.state.primary) {
                Log.error('Mongo topology lost connection to primary!', {
                    state: data.state,
                    numNoPrimaries: numNoPrimaries
                });
                if (!data.state.secondaries || data.state.secondaries.length < 1) {
                    Log.fatal('Lost connection to all mongo instances!');
                    return;
                }
                if (numNoPrimaries++ >= 1) {
                    Log.fatal('Lost connection to all mongo instances!');
                }
            }
        }.bind(this));
        topology.on('left', function(serverType, server) {
            //so server.name should be a getter to get the serverDetails
            //see: https://github.com/christkv/mongodb-core/blob/master/lib/topologies/server.js#L556
            Log.warn('Lost connection to mongo server', {type: server.name});
        }.bind(this));
    }

    server = new WebSocketServer({port: flags.get('port'), host: flags.get('ip')}, advertiseToSkyAPI);
    server.on('connection', function(ws) {
        var ip = ws._socket.remoteAddress;
        Log.debug('New ws connection', {ip: ip});

        ws.on('message', function(message) {
            Log.debug('New ws message', {message: message, ip: ip});
            try {
                handleMessage(localDB, ws, message);
            } catch (e) {
                Log.error('Error handling message', {error: e, ip: ip});
            }
        });
        ws.on('error', function(err) {
            Log.error('Error from ws', {err: err, ip: ip});
            var i = allClients.indexOf(ws);
            if (i !== -1) {
                allClients.splice(i, 1);
            }
        });
        ws.on('close', function() {
            Log.debug('Close from ws', {ip: ip});
            var i = allClients.indexOf(ws);
            if (i !== -1) {
                allClients.splice(i, 1);
            }
        });
    });

    if (flags.get('clean')) {
        setInterval(cleanupDeadReplicas.bind(null, localDB), 60000);
    }
});